tests/core/pthread/sync_to_async.cpp

```diff
@@ -0,0 +1,152 @@
+#include <assert.h>
+#include <emscripten.h>
+#include <pthread.h>
+
+#include <thread>
+#include <functional>
+#include <iostream>
+
+// Helper class for generic sync-to-async conversion. doWork() lets you provide
+// some code to be run, which is done on a helper thread. The thread can do
+// async operations, and it has a callback to indicate when it finishes working.
+// The calling thread mainwhile blocks as it waits.
+//
+// This can be useful if you are a location where blocking is possible (like a
+// thread, or when using PROXY_TO_PTHREAD), but you have code that is hard to
+// refactor to be async, but that requires some async operation (like waiting
+// for a JS event).
```

Comment:
Hmm, if it's waiting for a JS event, wouldn't it be more natural for such waiting code to be executed on the main thread instead of another helper thread instead?

In which case we're coming to something like https://github.com/emscripten-core/emscripten/issues/9709 - execute a piece of async JS and wait until it finishes.

Reply:
This is trying to help the use case where the current thread must run synchronously - for example, if the codebase cannot be rewritten to be async - and where Asyncify is not being run (because of overhead, build time, or the need for debug info, for example). You're right that not all things can be run on another thread, but at least some can, like a network fetch.